Optimize SVGRenderSupport::layoutDifferentRootIfNeeded()

smfr opened the pull request:
#### 82682c1a0a673dc7b4ae28fef1ee9c8a3209d4fd
<pre>
Optimize SVGRenderSupport::layoutDifferentRootIfNeeded()
<a href="https://bugs.webkit.org/show_bug.cgi?id=267014">https://bugs.webkit.org/show_bug.cgi?id=267014</a>
<a href="https://rdar.apple.com/120392477">rdar://120392477</a>

Reviewed by Said Abou-Hallawa.

SVGRenderSupport::layoutDifferentRootIfNeeded() shows up on profiles of the Suits
MotionMark subtest.

We only need to call findTreeRootObject() on the renderer if the there are any  clipper, masker,
filters etc are present, so pass the RenderElement which is the client of the SVG resources,
and lazily call findTreeRootObject() on it when necessary.

Also check m_clipperFilterMaskerData and m_markerData before calling the lambda on their
contents.

* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::layoutDifferentRootIfNeeded):
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::layoutDifferentRootIfNeeded):
* Source/WebCore/rendering/svg/SVGResources.h:

Canonical link: <a href="https://commits.webkit.org/272696@main">https://commits.webkit.org/272696@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d390658fba040542237f2a30de4c46dcb3842691

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32665 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11418 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34519 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35237 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29522 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33589 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13785 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8605 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29009 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33104 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9637 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29175 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8388 "Found 1 new API test failure: /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/service-worker-registrations (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8531 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29123 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36573 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29699 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29537 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34643 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8644 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6601 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32507 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/replaced-objects (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10326 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7601 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9258 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9248 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/svg/SVGResources.cpp:
```diff
@@ -307,22 +307,38 @@ std::unique_ptr<SVGResources> SVGResources::buildCachedResources(const RenderEle
-    layoutDifferentRootIfNeeded(markerEnd());
+
+    if (m_clipperFilterMaskerData) {
+        layoutDifferentRootIfNeeded(clipper());
```
This will save another check for `m_clipperFilterMaskerData` in `SVGResources::clipper()`

```
layoutDifferentRootIfNeeded(m_clipperFilterMaskerData->clipper.get());
```

shallawa commented on Source/WebCore/rendering/svg/SVGResources.cpp:
```diff
@@ -307,22 +307,38 @@ std::unique_ptr<SVGResources> SVGResources::buildCachedResources(const RenderEle
+        if (!clientSVGRoot)
+            clientSVGRoot = SVGRenderSupport::findTreeRootObject(resourcesClient);
+
+        if (clientSVGRoot == root)
```
I think this should be either `(clientSVGRoot == svgRoot)` or `(clientRoot == root)`

shallawa commented on Source/WebCore/rendering/svg/SVGResources.cpp:
```diff
@@ -307,22 +307,38 @@ std::unique_ptr<SVGResources> SVGResources::buildCachedResources(const RenderEle
 {
+    const LegacyRenderSVGRoot* clientSVGRoot = nullptr;
+
     auto layoutDifferentRootIfNeeded = [&](RenderElement* container) {
```
Should not the type be `LegacyRenderSVGResourceContainer` instead of `RenderElement`? And should not the name be `resourceContainer` or `resource` instead of the generic name `container`?

nikolaszimmermann reviewed (commented):
Thanks Simon, note that the whole mechanism is obsoleted in LBSE - still good to have the cleanup for legacy.

smfr committed:
Optimize SVGRenderSupport::layoutDifferentRootIfNeeded()
https://bugs.webkit.org/show_bug.cgi?id=267014
rdar://120392477

Reviewed by Said Abou-Hallawa.

SVGRenderSupport::layoutDifferentRootIfNeeded() shows up on profiles of the Suits
MotionMark subtest.

We only need to call findTreeRootObject() on the renderer if the there are any  clipper, masker,
filters etc are present, so pass the RenderElement which is the client of the SVG resources,
and lazily call findTreeRootObject() on it when necessary.

Also check m_clipperFilterMaskerData and m_markerData before calling the lambda on their
contents.

* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::layoutDifferentRootIfNeeded):
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::layoutDifferentRootIfNeeded):
* Source/WebCore/rendering/svg/SVGResources.h:

Canonical link: https://commits.webkit.org/272696@main
